sycl/doc/CompilerAndRuntimeDesign.md

```diff
@@ -787,18 +787,59 @@ The SPIR-V specific functions are implemented in for the SYCL host device here:
 
 ### Address spaces handling
 
-SYCL 1.2.1 language defines several address spaces where data can reside, the
-same as OpenCL - global, local, private and constant. From the spec: "In OpenCL
-C, these address spaces are manually specified using OpenCL-specific keywords.
-In SYCL, the device compiler is expected to auto-deduce the address space for
-pointers in common situations of pointer usage. However, there are situations
-where auto-deduction is not possible".
-
-We believe that requirement for the compiler to automatically deduce address
-spaces is too strong. Instead the following approach will be implemented in the
-compiler:
+SYCL-1.2.1 specification uses C++ templates to represent pointers to disjoint
+memory regions on an accelerator to enable compilation with standard C++
+toolchain and SYCL compiler toolchain.
+
+For instance:
+
+``` C++
+// check that SYCL mode is ON and we can use non-standard annotations
+#if defined(__SYCL_DEVICE_ONLY__)
+// GPU/accelerator implementation
+template <typename T, address_space AS> class multi_ptr {
+  // GetAnnotatedPointer<T, global>::type == "__attribute__((opencl_global)) T"
+  // See sycl/include/CL/sycl/access/access.hpp for more details
+  using pointer_t = typename GetAnnotatedPointer<T, AS>::type *;
+ 
+  pointer_t data;
+  public:
+  pointer_t get_pointer() { return data; }
+}
+#else
+// CPU/host implementation
+template <typename T, address_space AS> class multi_ptr {
+  T *data; // regular unannotated pointer
+  public:
+  T *get_pointer() { return data; }
+}
+#endif
+```
 
-*TBD*
+Depending on the compiler mode `multi_ptr` will either annotate internal data
+with address space attribute or not.
+
+To utilize existing clang's functionality, we re-use following OpenCL address
+space attributes in SYCL mode: `__attribute__((opencl_global))`,
+`__attribute__((opencl_global_host))`, `__attribute__((opencl_global_device))`,
+`__attribute__((opencl_local))`, `__attribute__((opencl_private))` and
+`__attribute__((opencl_constant))`.
+
+The main address space semantic difference of SYCL mode from OpenCL is that
+SYCL doesn't assign OpenCL generic address space to a declaration w/o explicit
+address space attribute. Similar to other single-source C++-based GPU
+programming modes like OpenMP/CUDA/HIP, SYCL keeps "default" address space for
+such declarations and defines pointer casts between OpenCL address space
+attributes and default address space the same way OpenCL specification defines
+them for generic address space. This design has two important features: keeps
+the type system consistent between host and device compilers on the one hand and
+enable tools for emitting device code aligned with SPIR memory model (and other
+GPU targets).
+
```

Comment:
```suggestion
address space attribute. Similar to other single-source C++-based GPU
programming modes like OpenMP/CUDA/HIP, SYCL uses
clang's "default" address space for types with no address space attributes.
During the lowering to LLVM IR, the default address space is mapped to
the SPIR generic address space. Declarations are assigned to the
relevant memory region depending on their declaration context and
pointers to them are cast to generic. This design has two important features: keeps
the type system consistent with C++ on one hand and enable tools for emitting device code
aligned with SPIR memory model (and other GPU targets).

```

This is really a suggestion, I tried here to 1) move away from OpenCL (use SPIR instead as this is the target) 2) emphasis that the address space assignment happens during lowering to LLVM IR rather than Sema.